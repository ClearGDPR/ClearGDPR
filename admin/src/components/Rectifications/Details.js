import React from 'react';
import PropTypes from 'prop-types';
import lodash from 'lodash';

import { rectificationDetailsType } from 'types';

import Loader from 'components/core/cards/dashboard/Loader';
import { PrimaryButton } from 'components/core/Common/Buttons/Buttons';

const Details = ({ rectification, onApprove, isLoading }) => {
  function renderObject(obj, level = 0) {
    if (!obj) return;

    return Object.keys(obj).map((k, index) => (
      <React.Fragment key={index}>
        {level ? <br /> : ''}
        {[...Array(level)].map((value, arrayIndex) => (
          <React.Fragment key={arrayIndex}>&nbsp;&nbsp;</React.Fragment>
        ))}
        {k}:&nbsp;
        {lodash.isObject(obj[k]) ? renderObject(obj[k], level + 1) : obj[k]}
        {!level ? <br /> : ''}
      </React.Fragment>
    ));
  }

  function renderDetails() {
    return (
      <React.Fragment>
        <p>
          <strong>Created at:</strong> {rectification.created_at}
        </p>
        <p>
          <strong>Current data:</strong>
        </p>
        <p>{renderObject(rectification.currentData)}</p>
        <p>
          <strong>Updates:</strong>
        </p>
        <p>{renderObject(rectification.updates)}</p>
        <p>Status: {rectification.status}</p>
        {rectification.status !== 'APPROVED' && (
          <PrimaryButton
            type="button"
            onClick={e => {
              e.preventDefault();
              onApprove();
            }}
            text="Approve"
          />
        )}
      </React.Fragment>
    );
  }
  return isLoading ? <Loader /> : renderDetails();
};

Details.propTypes = {
  rectification: rectificationDetailsType.isRequired,
  onApprove: PropTypes.func.isRequired,
  isLoading: PropTypes.bool
};

export default Details;
