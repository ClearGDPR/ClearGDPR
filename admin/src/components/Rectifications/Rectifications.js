import React from 'react';
import PropTypes from 'prop-types';

import { Tab, Tabs, TabList, TabPanel } from 'react-tabs';
import 'react-tabs/style/react-tabs.css';
import 'theme/Tabs.css';

import Table from 'components/core/Common/Table/Table';
import Loader from 'components/core/cards/dashboard/Loader';
import Card from 'components/core/cards/dashboard/Card';
import Paginate from 'components/core/Paginate';
import ActionBar from 'components/core/Common/Bars/ActionBar';

const actions = ({ onDetailsClick }) => [
  {
    label: 'Details',
    onClick: (e, item) => {
      e.preventDefault();
      onDetailsClick(item.id);
    }
  }
];

const columns = ({ showStatus } = { showStatus: false }) => ({
  id: {
    title: 'ID'
  },
  created_at: {
    title: 'Created at'
  },
  request_reason: {
    title: 'Reason'
  },
  status: {
    title: 'Status',
    filter: () => showStatus
  }
});

const Rectifications = ({
  isLoading,
  tabs,
  selectedTab,
  onTabSelect,
  pageCount,
  currentPage,
  onPageSelected,
  onDetailsClick,
  data
}) => {
  function handlePageClick(page) {
    onPageSelected(page);
  }

  function renderPagination() {
    return (
      <Paginate
        pageCount={pageCount}
        initialPage={currentPage - 1}
        onPageChange={({ selected }) => handlePageClick(selected + 1)}
        disableInitialCallback={true}
      />
    );
  }

  function renderPendingRequests() {
    return (
      <React.Fragment>
        <div className="content">
          {data.length > 0 ? (
            <Table rows={data} columns={columns()} actions={actions({ onDetailsClick })} />
          ) : (
            <i>No pending requests registered</i>
          )}
        </div>
        <div className="content">{renderPagination()}</div>
      </React.Fragment>
    );
  }

  function renderRequestsArchive() {
    return (
      <React.Fragment>
        <div className="content">
          {data.length > 0 ? (
            <Table
              rows={data}
              columns={columns({ showStatus: true })}
              actions={actions({ onDetailsClick })}
            />
          ) : (
            <i>No requests archieved</i>
          )}
        </div>
        <div className="content">{renderPagination()}</div>
      </React.Fragment>
    );
  }

  function renderLoader() {
    return (
      <div className="content">
        <Loader />
      </div>
    );
  }

  return (
    <section className="cards">
      <ActionBar
        title="Rectification requests"
        desc="Manage requests from users who want to execute their right to rectify the
            data stored about them."
      />
      <Tabs selectedIndex={selectedTab} onSelect={onTabSelect}>
        <TabList>{tabs.map((value, index) => <Tab key={index}>{value}</Tab>)}</TabList>
        <TabPanel>
          <div className="row">
            <Card cols={8}>{!isLoading ? renderPendingRequests() : renderLoader()}</Card>
          </div>
        </TabPanel>
        <TabPanel>
          <div className="row">
            <Card cols={8}>{!isLoading ? renderRequestsArchive() : renderLoader()}</Card>
          </div>
        </TabPanel>
      </Tabs>
    </section>
  );
};

Rectifications.propTypes = {
  isLoading: PropTypes.bool,
  tabs: PropTypes.array.isRequired,
  selectedTab: PropTypes.number.isRequired,
  onTabSelect: PropTypes.func.isRequired,
  pageCount: PropTypes.number.isRequired,
  currentPage: PropTypes.number.isRequired,
  onPageSelected: PropTypes.func.isRequired,
  onDetailsClick: PropTypes.func.isRequired,
  data: PropTypes.array.isRequired
};

export default Rectifications;
