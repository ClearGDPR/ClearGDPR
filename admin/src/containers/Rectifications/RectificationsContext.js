import React, { Component, createContext } from 'react';
import PropTypes from 'prop-types';
import config from 'config';
import internalFetch from 'helpers/internal-fetch';
import { toast } from 'react-toastify';
import { format } from 'date-fns';

const RectificationsContext = createContext({
  pendingRectifications: {},
  processedRectifications: {},
  fetchPendingRectifications: () => {},
  fetchProcessedRectifications: () => {},
  fetchAllRectifications: () => {},
  approveRectification: () => {},
  isLoading: false
});

export class RectificationsProvider extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.object,
      PropTypes.arrayOf(PropTypes.element)
    ])
  };

  fetchPendingRectifications = async (page = 1) => {
    await this._fetchRectifications(true, false, page);
  };

  fetchProcessedRectifications = async (page = 1) => {
    await this._fetchRectifications(false, true, page);
  };

  fetchAllRectifications = async () => {
    await this._fetchRectifications();
  };

  approveRectification = async id => {
    await Promise.resolve();
  };

  state = {
    pendingRectifications: { paging: { current: 1 } },
    processedRectifications: { paging: { current: 1 } },
    fetchPendingRectifications: this.fetchPendingRectifications,
    fetchProcessedRectifications: this.fetchProcessedRectifications,
    fetchAllRectifications: this.fetchAllRectifications,
    approveRectification: this.approveRectification,
    isLoading: false
  };

  async _getPendingRectifications(page = 1) {
    return this._mapRectificationResults(
      await internalFetch(
        `${config.API_URL}/api/management/subjects/rectification-requests/list?page=${page}`
      )
    );
  }

  async _getProcessedRectifications(page = 1) {
    return this._mapRectificationResults(
      await internalFetch(
        `${config.API_URL}/api/management/subjects/rectification-requests/archive?page=${page}`
      )
    );
  }

  _mapRectificationResults(result) {
    const mappedData = result.data.map(v => {
      let mapped = {
        id: v.id,
        created_at: format(new Date(v.created_at), 'DD/MM/YYYY h:mma'),
        request_reason: v.request_reason
      };

      if (v.status) mapped.status = v.status;
      return mapped;
    });
    return {
      ...result,
      data: mappedData
    };
  }

  setLoading(loading) {
    this.setState({
      isLoading: loading
    });
  }

  async cancelLoadingAndReject(e) {
    this.setLoading(false);
    toast.error(`An error occurred: ${e.message}`);
  }

  async _fetchRectifications(pending = true, processed = true, page) {
    this.setLoading(true);

    try {
      const pendingPage = page || this.state.pendingRectifications.paging.current;
      const pendingRectifications = pending
        ? await this._getPendingRectifications(pendingPage)
        : this.state.pendingRectifications;
      const processedRectifications = processed
        ? await this._getProcessedRectifications(
            page || this.state.processedRectifications.paging.current
          )
        : this.state.processedRectifications;
      this.setState({
        pendingRectifications,
        processedRectifications,
        isLoading: false
      });
    } catch (e) {
      await this.cancelLoadingAndReject(e);
    }
  }

  render() {
    return (
      <RectificationsContext.Provider value={this.state}>
        {this.props.children}
      </RectificationsContext.Provider>
    );
  }
}

export const RectificationsConsumer = RectificationsContext.Consumer;
