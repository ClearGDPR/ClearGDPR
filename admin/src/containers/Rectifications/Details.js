import React from 'react';
import PropTypes from 'prop-types';
import format from 'helpers/date-formatter';

import config from 'config';
import internalFetch from 'helpers/internal-fetch';

import { RectificationsConsumer } from './RectificationsContext';
import { PanelConsumer } from 'containers/MainLayout/PanelContext';
import Details from 'components/Rectifications/Details';

export class DetailsContainer extends React.Component {
  static propTypes = {
    approveRectification: PropTypes.func,
    rectificationId: PropTypes.number,
    closePanel: PropTypes.func,
    isLoading: PropTypes.bool
  };

  state = {
    rectification: {}
  };

  componentDidMount() {
    this.setState({
      isLoading: true
    });

    internalFetch(
      `${config.API_URL}/api/management/subjects/rectification-requests/${
        this.props.rectificationId
      }`
    )
      .then(r =>
        this.setState({
          rectification: {
            created_at: format(r.createdAt),
            currentData: r.currentData,
            updates: r.updates,
            status: r.status
          },
          isLoading: false
        })
      )
      .catch(() => this.setState({ isLoading: false }));
  }

  get isLoading() {
    return this.state.isLoading || this.props.isLoading;
  }

  onApprove = () => {
    this.setState({
      isLoading: true
    });

    this.props
      .approveRectification(this.props.rectificationId)
      .then(() => this.props.closePanel())
      .catch(() => {
        this.setState({
          isLoading: false
        });
      });
  };

  render() {
    return (
      <Details
        rectification={this.state.rectification}
        onApprove={this.onApprove}
        isLoading={this.isLoading}
      />
    );
  }
}

export default props => (
  <PanelConsumer>
    {({ closePanel }) => (
      <RectificationsConsumer>
        {({ approveRectification, isLoading }) => (
          <DetailsContainer
            {...props}
            isLoading={isLoading}
            approveRectification={approveRectification}
            closePanel={closePanel}
          />
        )}
      </RectificationsConsumer>
    )}
  </PanelConsumer>
);
