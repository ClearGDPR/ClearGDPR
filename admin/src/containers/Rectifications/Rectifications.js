import React from 'react';
import PropTypes from 'prop-types';

import { RectificationsConsumer } from 'containers/Rectifications/RectificationsContext';
import { PanelConsumer } from 'containers/MainLayout/PanelContext';

import Rectifications from 'components/Rectifications/Rectifications';
import Details from 'components/Rectifications/Details';

export class RectificationsContainer extends React.Component {
  static propTypes = {
    pendingRectifications: PropTypes.object.isRequired,
    processedRectifications: PropTypes.object.isRequired,
    fetchPendingRectifications: PropTypes.func.isRequired,
    fetchProcessedRectifications: PropTypes.func.isRequired,
    fetchAllRectifications: PropTypes.func.isRequired,
    isLoading: PropTypes.bool.isRequired,
    openPanel: PropTypes.func.isRequired
  };

  state = {
    selectedTab: 0,
    tabs: ['Pending requests', 'Requests archive']
  };

  get pendingTabActive() {
    return this.state.selectedTab === 0;
  }

  onTabSelect(tabIndex) {
    this.setState({
      selectedTab: tabIndex
    });
  }

  getRectificationsData() {
    if (this.pendingTabActive) {
      return this.props.pendingRectifications;
    } else {
      return this.props.processedRectifications;
    }
  }

  getData() {
    return this.getRectificationsData().data || [];
  }

  componentDidMount() {
    this.props.fetchAllRectifications();
  }

  getPageCount() {
    const paging = this.getRectificationsData().paging;
    return paging ? paging.total : 1;
  }

  getCurrentPage() {
    const paging = this.getRectificationsData().paging;
    return paging ? paging.current : 1;
  }

  onPageSelected(page) {
    if (this.pendingTabActive) {
      this.props.fetchPendingRectifications(page);
    } else {
      this.props.fetchProcessedRectifications(page);
    }
  }

  onDetailsClick(id) {
    this.props.openPanel(Details, 'Rectification details', { rectificationId: id });
  }

  render() {
    return (
      <Rectifications
        tabs={this.state.tabs}
        selectedTab={this.state.selectedTab}
        onTabSelect={this.onTabSelect.bind(this)}
        pageCount={this.getPageCount()}
        currentPage={this.getCurrentPage()}
        data={this.getData()}
        isLoading={this.props.isLoading}
        onPageSelected={this.onPageSelected.bind(this)}
        onDetailsClick={this.onDetailsClick.bind(this)}
      />
    );
  }
}

export default props => (
  <PanelConsumer>
    {({ openPanel }) => (
      <RectificationsConsumer>
        {({
          pendingRectifications,
          processedRectifications,
          fetchPendingRectifications,
          fetchProcessedRectifications,
          fetchAllRectifications,
          isLoading
        }) => (
          <RectificationsContainer
            {...props}
            pendingRectifications={pendingRectifications}
            processedRectifications={processedRectifications}
            fetchPendingRectifications={fetchPendingRectifications}
            fetchProcessedRectifications={fetchProcessedRectifications}
            fetchAllRectifications={fetchAllRectifications}
            isLoading={isLoading}
          />
        )}
      </RectificationsConsumer>
    )}
  </PanelConsumer>
);
