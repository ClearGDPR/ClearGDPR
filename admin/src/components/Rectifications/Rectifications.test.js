import React from 'react';
import { shallow } from 'enzyme';
import Rectifications from './Rectifications';

const rectifications = {
  data: [
    {
      id: 4,
      request_reason: 'The data was incorrect.',
      created_at: '02/07/2018 9:31pm',
      status: 'APPROVED'
    },
    {
      id: 5,
      request_reason: 'The data was incorrect two.',
      created_at: '02/07/2018 9:35pm',
      status: 'DISAPPROVED'
    },
    {
      id: 6,
      request_reason: 'The data was incorrect two three.',
      created_at: '02/07/2018 9:31pm',
      status: 'APPROVED'
    }
  ],
  paging: {
    current: 1,
    total: 2
  }
};

const setup = propOverrides => {
  const props = {
    data: rectifications.data,
    tabs: ['tab 1', 'tab 2'],
    selectedTab: 0,
    onTabSelect: jest.fn(),
    pageCount: rectifications.paging.total,
    currentPage: rectifications.paging.current,
    onPageSelected: jest.fn(),
    isLoading: false,
    onDetailsClick: jest.fn(),
    ...propOverrides
  };

  const component = shallow(<Rectifications {...props} />);

  return { props, component };
};

describe('(Component) Rectifications', () => {
  it('should render when not loading correctly', () => {
    const { component } = setup();
    expect(component).toMatchSnapshot();
  });

  it('should render when loading correctly', () => {
    const component = setup({ isLoading: true });
    expect(component).toMatchSnapshot();
  });
});
