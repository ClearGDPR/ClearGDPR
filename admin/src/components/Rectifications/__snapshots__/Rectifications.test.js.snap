// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Rectifications should render when loading correctly 1`] = `
Object {
  "component": <section
    className="cards"
  >
    <ActionBar
      desc="Here you can manage requests from users who want to execute their right to rectify the data stored about them."
      title="Rectification requests"
    />
    <Tabs
      defaultFocus={false}
      defaultIndex={null}
      forceRenderTabPanel={false}
      onSelect={[MockFunction]}
      selectedIndex={0}
    >
      <TabList
        className="react-tabs__tab-list"
      >
        <Tab
          className="react-tabs__tab"
          disabledClassName="react-tabs__tab--disabled"
          focus={false}
          id={null}
          key="0"
          panelId={null}
          selected={false}
          selectedClassName="react-tabs__tab--selected"
        >
          tab 1
        </Tab>
        <Tab
          className="react-tabs__tab"
          disabledClassName="react-tabs__tab--disabled"
          focus={false}
          id={null}
          key="1"
          panelId={null}
          selected={false}
          selectedClassName="react-tabs__tab--selected"
        >
          tab 2
        </Tab>
      </TabList>
      <TabPanel
        className="react-tabs__tab-panel"
        forceRender={false}
        selectedClassName="react-tabs__tab-panel--selected"
      >
        <div
          className="row"
        >
          <Card
            cols={8}
          >
            <div
              className="content"
            >
              <Loader />
            </div>
          </Card>
        </div>
      </TabPanel>
      <TabPanel
        className="react-tabs__tab-panel"
        forceRender={false}
        selectedClassName="react-tabs__tab-panel--selected"
      >
        <div
          className="row"
        >
          <Card
            cols={8}
          >
            <div
              className="content"
            >
              <Loader />
            </div>
          </Card>
        </div>
      </TabPanel>
    </Tabs>
  </section>,
  "props": Object {
    "currentPage": 1,
    "data": Array [
      Object {
        "created_at": "02/07/2018 9:31pm",
        "id": 4,
        "request_reason": "The data was incorrect.",
        "status": "APPROVED",
      },
      Object {
        "created_at": "02/07/2018 9:35pm",
        "id": 5,
        "request_reason": "The data was incorrect two.",
        "status": "DISAPPROVED",
      },
      Object {
        "created_at": "02/07/2018 9:31pm",
        "id": 6,
        "request_reason": "The data was incorrect two three.",
        "status": "APPROVED",
      },
    ],
    "isLoading": true,
    "onDetailsClick": [MockFunction],
    "onPageSelected": [MockFunction],
    "onTabSelect": [MockFunction],
    "pageCount": 2,
    "selectedTab": 0,
    "tabs": Array [
      "tab 1",
      "tab 2",
    ],
  },
}
`;

exports[`(Component) Rectifications should render when not loading correctly 1`] = `
<section
  className="cards"
>
  <ActionBar
    desc="Here you can manage requests from users who want to execute their right to rectify the data stored about them."
    title="Rectification requests"
  />
  <Tabs
    defaultFocus={false}
    defaultIndex={null}
    forceRenderTabPanel={false}
    onSelect={[MockFunction]}
    selectedIndex={0}
  >
    <TabList
      className="react-tabs__tab-list"
    >
      <Tab
        className="react-tabs__tab"
        disabledClassName="react-tabs__tab--disabled"
        focus={false}
        id={null}
        key="0"
        panelId={null}
        selected={false}
        selectedClassName="react-tabs__tab--selected"
      >
        tab 1
      </Tab>
      <Tab
        className="react-tabs__tab"
        disabledClassName="react-tabs__tab--disabled"
        focus={false}
        id={null}
        key="1"
        panelId={null}
        selected={false}
        selectedClassName="react-tabs__tab--selected"
      >
        tab 2
      </Tab>
    </TabList>
    <TabPanel
      className="react-tabs__tab-panel"
      forceRender={false}
      selectedClassName="react-tabs__tab-panel--selected"
    >
      <div
        className="row"
      >
        <Card
          cols={8}
        >
          <React.Fragment>
            <div
              className="content"
            >
              <Table
                actions={
                  Array [
                    Object {
                      "label": "Details",
                      "onClick": [Function],
                    },
                  ]
                }
                columns={
                  Object {
                    "created_at": Object {
                      "title": "Created at",
                    },
                    "id": Object {
                      "title": "ID",
                    },
                    "request_reason": Object {
                      "title": "Reason",
                    },
                    "status": Object {
                      "filter": [Function],
                      "title": "Status",
                    },
                  }
                }
                rows={
                  Array [
                    Object {
                      "created_at": "02/07/2018 9:31pm",
                      "id": 4,
                      "request_reason": "The data was incorrect.",
                      "status": "APPROVED",
                    },
                    Object {
                      "created_at": "02/07/2018 9:35pm",
                      "id": 5,
                      "request_reason": "The data was incorrect two.",
                      "status": "DISAPPROVED",
                    },
                    Object {
                      "created_at": "02/07/2018 9:31pm",
                      "id": 6,
                      "request_reason": "The data was incorrect two three.",
                      "status": "APPROVED",
                    },
                  ]
                }
              />
            </div>
            <div
              className="content"
            >
              <Paginate
                disableInitialCallback={true}
                initialPage={0}
                onPageChange={[Function]}
                pageCount={2}
              />
            </div>
          </React.Fragment>
        </Card>
      </div>
    </TabPanel>
    <TabPanel
      className="react-tabs__tab-panel"
      forceRender={false}
      selectedClassName="react-tabs__tab-panel--selected"
    >
      <div
        className="row"
      >
        <Card
          cols={8}
        >
          <React.Fragment>
            <div
              className="content"
            >
              <Table
                actions={
                  Array [
                    Object {
                      "label": "Details",
                      "onClick": [Function],
                    },
                  ]
                }
                columns={
                  Object {
                    "created_at": Object {
                      "title": "Created at",
                    },
                    "id": Object {
                      "title": "ID",
                    },
                    "request_reason": Object {
                      "title": "Reason",
                    },
                    "status": Object {
                      "filter": [Function],
                      "title": "Status",
                    },
                  }
                }
                rows={
                  Array [
                    Object {
                      "created_at": "02/07/2018 9:31pm",
                      "id": 4,
                      "request_reason": "The data was incorrect.",
                      "status": "APPROVED",
                    },
                    Object {
                      "created_at": "02/07/2018 9:35pm",
                      "id": 5,
                      "request_reason": "The data was incorrect two.",
                      "status": "DISAPPROVED",
                    },
                    Object {
                      "created_at": "02/07/2018 9:31pm",
                      "id": 6,
                      "request_reason": "The data was incorrect two three.",
                      "status": "APPROVED",
                    },
                  ]
                }
              />
            </div>
            <div
              className="content"
            >
              <Paginate
                disableInitialCallback={true}
                initialPage={0}
                onPageChange={[Function]}
                pageCount={2}
              />
            </div>
          </React.Fragment>
        </Card>
      </div>
    </TabPanel>
  </Tabs>
</section>
`;
